rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbDeviceStateNode.java

```diff
@@ -0,0 +1,187 @@
+            TbMsgType.CONNECT_EVENT, this::sendDeviceConnectMsg,
+            TbMsgType.ACTIVITY_EVENT, this::sendDeviceActivityMsg,
+            TbMsgType.DISCONNECT_EVENT, this::sendDeviceDisconnectMsg,
+            TbMsgType.INACTIVITY_EVENT, this::sendDeviceInactivityMsg
+    );
+
+    private interface ConnectivityEvent {
+
+        void sendEvent(TbContext ctx, TbMsg msg);
+
+    }
+
+    private TbMsgType event;
+
+    @Override
+    public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
+        TbMsgType event = TbNodeUtils.convert(configuration, TbDeviceStateNodeConfiguration.class).getEvent();
+        if (event == null) {
+            throw new TbNodeException("Event cannot be null!", true);
+        }
+        if (!SUPPORTED_EVENTS.containsKey(event)) {
+            throw new TbNodeException("Unsupported event: " + event, true);
+        }
+        this.event = event;
+    }
+
+    @Override
+    public void onMsg(TbContext ctx, TbMsg msg) {
+        var originator = msg.getOriginator();
+        if (!ctx.isLocalEntity(originator)) {
+            log.warn("[{}][device-state-node] Received message from non-local entity [{}]!", ctx.getSelfId(), originator);
+            return;
+        }
+        if (!EntityType.DEVICE.equals(originator.getEntityType())) {
+            ctx.tellSuccess(msg);
+            return;
+        }
+        SUPPORTED_EVENTS.get(event).sendEvent(ctx, msg);
+        ctx.tellSuccess(msg);
+    }
+
+    private void sendDeviceConnectMsg(TbContext ctx, TbMsg msg) {
+        var tenantUuid = ctx.getTenantId().getId();
+        var deviceUuid = msg.getOriginator().getId();
+        var deviceConnectMsg = TransportProtos.DeviceConnectProto.newBuilder()
+                .setTenantIdMSB(tenantUuid.getMostSignificantBits())
+                .setTenantIdLSB(tenantUuid.getLeastSignificantBits())
+                .setDeviceIdMSB(deviceUuid.getMostSignificantBits())
+                .setDeviceIdLSB(deviceUuid.getLeastSignificantBits())
+                .build();
+        var toCoreMsg = TransportProtos.ToCoreMsg.newBuilder()
+                .setDeviceConnectMsg(deviceConnectMsg)
+                .build();
+        ctx.getClusterService().pushMsgToCore(
+                ctx.getTenantId(), msg.getOriginator(), toCoreMsg, getMsgProcessedCallback(ctx, msg)
+        );
+    }
+
+    private void sendDeviceActivityMsg(TbContext ctx, TbMsg msg) {
+        var tenantUuid = ctx.getTenantId().getId();
+        var deviceUuid = msg.getOriginator().getId();
+        var deviceActivityMsg = TransportProtos.DeviceActivityProto.newBuilder()
+                .setTenantIdMSB(tenantUuid.getMostSignificantBits())
+                .setTenantIdLSB(tenantUuid.getLeastSignificantBits())
+                .setDeviceIdMSB(deviceUuid.getMostSignificantBits())
+                .setDeviceIdLSB(deviceUuid.getLeastSignificantBits())
+                .setLastActivityTime(System.currentTimeMillis())
```

Comment:
Why the current wall clock? The activity event might come from the queue after a while

Reply:
Not sure what are you trying to say